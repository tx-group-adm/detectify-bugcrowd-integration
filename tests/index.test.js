const fetch = require('node-fetch');

const { Response } = jest.requireActual('node-fetch');
const { makeHeaders, signatureHeaders, getFindings } = require('../services/detectifyService');
const { getBounties, createSubmission, scoreCalculation } = require('../services/bugCrowdService');
const { errorHandling } = require('../utils/index');
const { submissionMock, findingsMock, bountiesMock, errorMock } = require('./mocks');

jest.mock('node-fetch');

// testing keys
const apiKey = '10840b0f938942feafb7186de74b9682';
const secretKey = '0vyTnawJRFn0Q9tWLTM188Olizc72JczHSXoIlsPQIc=';
const mockResponse = () => {
  const res = {};
  res.status = jest.fn().mockReturnValue(res);
  res.json = jest.fn().mockReturnValue(res);
  return res;
};

test('Checking for signature headers', () => {
  const signatureHeadersResult = {
    'X-Detectify-Signature': '6jpu6S4cQwEY4uLk+xELSe1RhajVJP0QEDpGWZ5T+U0=',
    'X-Detectify-Timestamp': 1519829567,
  };
  expect(signatureHeaders(apiKey, secretKey, 'GET', '/v2/domains/', 1519829567, null)).toStrictEqual(signatureHeadersResult);
});

test('scoreCalculation', () => {
  expect(scoreCalculation(2)).toBe(1);
  expect(scoreCalculation(4.4)).toBe(3);
  expect(scoreCalculation(7)).toBe(5);
  expect(scoreCalculation('null value')).toBe(NaN);
});

test('make header from signature', () => {
  const resultHeader = {
    'X-Detectify-Key': apiKey,
    'X-Detectify-Signature': '6jpu6S4cQwEY4uLk+xELSe1RhajVJP0QEDpGWZ5T+U0=',
    'X-Detectify-Timestamp': 1519829567,
  };
  expect(
    makeHeaders({
      apiKey,
      secretKey,
      method: 'GET',
      path: '/v2/domains/',
      timestamp: 1519829567,
      body: null,
    })
  ).toStrictEqual(resultHeader);
});

beforeEach(() => {
  fetch.mockClear();
});

test('getFindings', async () => {
  fetch.mockReturnValue(Promise.resolve(new Response(JSON.stringify(findingsMock))));

  const url = '/v2/findings/report/36df7ce57023ca2ftest/e31bd6160eaf1d212db348cc9test/66bcf6d0-4ca0-42ae-test/';
  try {
    const response = await getFindings(url);

    expect(fetch).toHaveBeenCalledTimes(1);
    expect(response).toMatchSnapshot();
    expect(response).toEqual(findingsMock);
  } catch (error) {
    errorHandling('error getting findings', 'Something went wrong', mockResponse());
  }
});

test('getFindings with error', async () => {
  fetch.mockReturnValue(Promise.resolve(new Response(JSON.stringify(errorMock))));

  const url = '/v2/findings/report/wrong/url';
  try {
    const response = await getFindings(url);

    expect(fetch).toHaveBeenCalledTimes(1);
    expect(response).toMatchSnapshot();
    expect(response).toEqual(findingsMock);
  } catch (error) {
    errorHandling(errorMock.message, errorMock, mockResponse());
  }
});

test('getBounties', async () => {
  fetch.mockReturnValue(Promise.resolve(new Response(JSON.stringify(bountiesMock))));

  const response = await getBounties();

  expect(fetch).toHaveBeenCalledTimes(1);
  expect(response).toMatchSnapshot();
  expect(response).toEqual(bountiesMock);
  expect(response).toHaveProperty('uuid');
  expect(fetch).toHaveBeenCalledWith('https://api.bugcrowd.com/bounties', {
    headers: {
      Accept: 'application/vnd.bugcrowd+json',
      Authorization: `Token ${process.env.BUG_CROWD_TOKEN}`,
      'Content-Type': 'application/json',
    },
    method: 'GET',
  });
});

test('getBounties error', async () => {
  fetch.mockReturnValue(Promise.resolve(new Response(JSON.stringify('Something went wrong!!!'))));
  try {
    const response = await getBounties();
    expect(fetch).toHaveBeenCalledTimes(1);
    expect(response).toMatchSnapshot();
    expect(response).toEqual(bountiesMock);
    expect(response).toHaveProperty('uuid');
  } catch (error) {
    errorHandling('Error getting bounties', "{'error':true, 'message': 'Something went wrong!!!'}", mockResponse());
  }
});

test('createSubmission', async () => {
  fetch.mockReturnValue(Promise.resolve(new Response(JSON.stringify(submissionMock))));
  const response = await createSubmission(findingsMock);

  expect(fetch).toHaveBeenCalledTimes(1);
  expect(response).toEqual(submissionMock);
  expect(fetch).toHaveBeenCalledWith(`https://api.bugcrowd.com/bounties/${process.env.BOUNTY_UUID}/submissions`, {
    headers: {
      Accept: 'application/vnd.bugcrowd+json',
      Authorization: `Token ${process.env.BUG_CROWD_TOKEN}`,
      'Content-Type': 'application/json',
    },
    method: 'POST',
    body: JSON.stringify({
      submission: {
        title: findingsMock.title,
        bug_url: findingsMock.found_at,
        description_markdown: findingsMock.definition.description,
        priority: scoreCalculation(findingsMock.score[0].score),
        substate: 'unresolved',
        extra_info_markdown: findingsMock.definition.risk,
        researcher_email: 'security@tx.group',
      },
    }),
  });
  expect(response).toMatchSnapshot();
});
