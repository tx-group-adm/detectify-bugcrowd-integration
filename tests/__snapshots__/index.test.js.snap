// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createSubmission 1`] = `
Object {
  "bounty": Object {
    "bounty_type": "ongoing",
    "code": "apitest-6dy5n",
    "custom_field_labels": Array [],
    "name": "api_test",
  },
  "bounty_brief_id": 14829,
  "bounty_code": "apitest-6dy5n",
  "bug_url": "https://login-igr.tagesanzeiger.ch/_next/static/chunks/commons.5459f0bd41ac674979bc.js.map",
  "caption": "JavaScript webpack sourcemap disclosure",
  "description_markdown": "Are you aware of that this file exists? This file may contain sensitive information.",
  "source": "api",
  "submitted_at": "2020-08-16T01:52:49.796Z",
  "substate": "unresolved",
  "title": "JavaScript webpack sourcemap disclosure",
  "uuid": "ac0ca949-4c0c-409e-8162-test",
}
`;

exports[`getBounties 1`] = `
Object {
  "bounty_type": "ongoing",
  "code": "apitest-6dy5n",
  "description_markdown": "Our company recognizes the importance of security, privacy and community, and values the input of hackers acting in good-faith to help us maintain a high standard for our users. This includes encouraging responsible vulnerability research and the disclosure of security vulnerabilities. ",
  "name": "api_test",
  "organization": Object {
    "name": "TX Group",
    "uuid": "7d844c76-6444-4b01-aecc-test",
  },
  "tagline": "Bug Bounty Program - Powered by Bugcrowd",
  "targets_overview_markdown": "Testing is only authorized on the targets listed as In-Scope. _Any domain/property of {company} not listed in the targets section is out of scope. This includes any/all subdomains not listed above._ If you believe you've identified a vulnerability on a system outside the scope, please reach out to support@bugcrowd.com before submitting. ----- ##Access: ###Credentials: ##Focus Areas: - one - two ----- ##Out-of-Scope - one - two ------ ##Safe Harbor: **When conducting vulnerability research according to this policy, we consider this research to be:** - Authorized in accordance with the Computer Fraud and Abuse Act (CFAA) (and/or similar state laws), and we will not initiate or support legal action against you for accidental, good faith violations of this policy; - Exempt from the Digital Millennium Copyright Act (DMCA), and we will not bring a claim against you for circumvention of technology controls; - Exempt from restrictions in our Terms & Conditions that would interfere with conducting security research, and we waive those restrictions on a limited basis for work done under this policy; and - Lawful, helpful to the overall security of the Internet, and conducted in good faith. - You are expected, as always, to comply with all applicable laws. _If at any time you have concerns or are uncertain whether your security research is consistent with this policy, please inquire via support@bugcrowd.com before going any further._ ",
  "uuid": "357a81ca-92b7-4b60-9548-test",
}
`;

exports[`getBounties error 1`] = `"Something went wrong!!!"`;

exports[`getFindings 1`] = `
Object {
  "definition": Object {
    "description": "Are you aware of that this file exists? This file may contain sensitive information.",
    "uuid": "d36ce0b7-9540-4cba-97c3-test",
  },
  "found_at": "https://login-igr.tagesanzeiger.ch/_next/static/chunks/commons.5459f0bd41ac674979bc.js.map",
  "report_token": "11ed472293aa5c466adc6afa3b0etest",
  "scan_profile_token": "36df7ce57023ca2fa70716test",
  "score": Array [
    Object {
      "score": 3.3,
      "vector": "AV:N/AC:L/Au:N/C:P/I:N/A:N/E:U/RL:OF/RC:UC",
      "version": "2.0",
    },
  ],
  "signature": "ca0f490a28802de0b1ef231816test",
  "timestamp": "2020-08-16T01:48:47Z",
  "title": "JavaScript webpack sourcemap disclosure",
  "url": "https://detectify.com/report/36df7ce57023ca2fa707160test/11ed472293aa5c466adc6afa3b0e7test/e26b149d-a124-44a6-a6c0-testS/",
  "uuid": "e26b149d-a124-44a6-a6c0-test",
}
`;

exports[`getFindings with error 1`] = `
Object {
  "code": 4051,
  "message": "Finding UUID is malformed",
  "more_info": "https://developer.detectify.com/#error-4051",
  "parameters": Array [
    "e5cf6af0-ceb5-4872-b823-110d5e89af80a",
  ],
}
`;
